### ISSUE TITLE
Cost of validator is prohibitive

### ISSUE THREAD

**Issue by cournape** [hynek/attrs#28]
I noticed that creating instances of attr objects was expensive:

```
# python >= 3.3, rough timing
import time

from attr import attributes, attr
from attr.validators import instance_of

@attributes
class Arch(object):
    name = attr("name")

@attributes
class ValidatedArch(object):
    name = attr("name", validator=instance_of(str))

def build_platform():
    return Arch("x86")

def timeit(func):
    times = []
    for i in range(10000):
        start = time.perf_counter()
        func()
        times.append(time.perf_counter() - start)
    return "Min cost across 10000 iterations: {:.2f} us".format(min(times) * 1e6)

print(timeit(lambda: Arch("x86")))
print(timeit(lambda: ValidatedArch("x86")))
```

I see a factor of ~ 50x (0.8us vs 34 us) on my machine. A quick profiling shows that almost all the difference comes from the deepcopy in the `attr._make.fields` function.

For non trivial classes with a few attributes, the construction cost quickly reaches the hundreds of us.

**Comment by hynek** [comment]
Hi, thanks for the detailed report!

I indeed haven’t cared much about creation performance until now and it seems like I was a bit too careful by using a public API that uses deep copy there.

Just iterating over `__attrs_attrs_` does this:

```
$ python t.py # before
Min cost across 10000 iterations: 0.55 us
Min cost across 10000 iterations: 31.03 us
$ python t.py # after
Min cost across 10000 iterations: 0.56 us
Min cost across 10000 iterations: 1.31 us
```

Since the `__attrs_attrs__` object is read-only after creation, it should be safe to not lock anything.

Opinions?

**Comment by cournape** [comment]
That certainly seems to fix the most glaring performance issues I noticed with attrs. If `__attrs_attrs__` is indeed read-only, that seems like a no-brainer ?

**Comment by cournape** [comment]
This issue is a bit critical for us, is there something more I can do to make this fix appear in a next release ?

**Comment by cournape** [comment]
(did not want to sound too pushy, if it is difficult to make it happen soon, we can fairly easily backport this in a bundled attr until a new official release)

**Comment by hynek** [comment]
If you’d have pinged the PR yesterday (GH doesn’t notify me of new pushes…), I might have made a quick release today but now I’m packing up for PyCon JP and won’t be back until next Wednesday.  I will try to find a minute to push it out tho.

**Comment by cournape** [comment]
Have fun at PyCon JP. Looks like this year's location is near tsukuji market, perfect for great sushi before the conference

**Comment by hynek** [comment]
Closed by #29
